### PR TITLE
Enabling matrix intermediate product simplification by default

### PR DESCRIPTION
**Pull request by tom-pytel**
#### References to other Issues or PRs
Implementing #18563

#### Brief description of what is fixed or changed
Enabling previously optional matrix intermediate product simplification by default since it seems to work well enough and to allow removal of `dotprodsimp` keyword arg in many matrix functions,

#### Release Notes

* matrices
  * Added intermediate matrix product simplification to control expression blowup during certain matrix operations.

### CODE REVIEW

**Comment by Pristine-Cat** [comment]
The tests will fail for docs but should in theory be fine for code as it was locally, but I don't test all the different configurations like Travis does so lets see.

**Comment by oscarbenjamin** [comment]
The commented geometry test hangs while trying to compute the minpoly here:
```
from sympy import *

minpoly(-S(497)/10 +
    (-175607206065265278663975578*sqrt(3745652738407155385007704061772062836444066)
    - 11101617483471454716584*sqrt(135224796172586744018128183165684231369061557693381)
    - 11191033528*sqrt(53763989816159429525176362970804231813094485229325771802373132829334557181)
    - 52787894*sqrt(939031677754234155852726812209396966309102909590089616712084247253381793874266)
    + 268217984113050831105496373604033955013*sqrt(405639795226)
    + 493031952994506001*sqrt(64122561702644519042349863936390022925453504764079220424826)
    + 1203794896762755972197548*sqrt(101693521550782291878002989407322636803510877426)
    + 304065484653645607594*sqrt(3671318908657008440503746493147223344418878917492544141)
    + 765229855267468398870*sqrt(2361810714906298822628782916193596395606230276564927466)
    + 450233845361953092755202061232170*sqrt(9233938046734998932847785072341)
    + 2202201959431354816517684012770622380006948*sqrt(630547164901) +
    5547966429911502647812643361890*sqrt(161278215511045609432874866293590026)
    + 6796630032876858542046645498857233390*sqrt(255775022850776494562626)
    + 3482205151600830212003049441707314361900680200000)/(1060000*sqrt(939031677754234155852726812209396966309102909590089616712084247253381793874266)
    + 224720000*sqrt(53763989816159429525176362970804231813094485229325771802373132829334557181)
    + 9920140080010000*sqrt(64122561702644519042349863936390022925453504764079220424826)
    + 222925387877984720000*sqrt(135224796172586744018128183165684231369061557693381)
    + 5734479977634180000*sqrt(3671318908657008440503746493147223344418878917492544141)
    + 6186432255656197934820000*sqrt(3745652738407155385007704061772062836444066)
    + 8314353298307300000*sqrt(2361810714906298822628782916193596395606230276564927466)
    + 44001968544258286480000*sqrt(101693521550782291878002989407322636803510877426)
    + 34598077337318951575033596144985271970000*sqrt(405639795226)
    + 8734410366439282997054067700000*sqrt(9233938046734998932847785072341)
    + 35787615129818861629958488626871981480000*sqrt(630547164901)
    + 78750830060908764094606460000*sqrt(161278215511045609432874866293590026)
    + 87084879438475523101972682336300000*sqrt(255775022850776494562626)
    + 70064489971847690382355119551455017342066000000))
```
This comes from
https://github.com/sympy/sympy/blob/a8a3a3b026cc55aa14010fc7cd7909806b6e116c/sympy/core/expr.py#L909
I don't think that minpoly should be used there.

**Comment by oscarbenjamin** [comment]
> I don't think that minpoly should be used there.

Removing it does give failures on master

**Comment by sylee957** [comment]
If some tests are look too artificial, I may allow skipping some.
But I'd just want to see how the benchmark looks like after passing the tests.

**Comment by Pristine-Cat** [comment]
> If some tests are look too artificial, I may allow skipping some.
> But I'd just want to see how the benchmark looks like after passing the tests.

I will fix the other tests and docs on my Linux env tomorrow to get this to pass the tests so it can be benchmarked properly. I will also push the sympy_benchmarks matrix benchmarks I created when testing dotprodsimp for more thorough tests.

**Comment by Pristine-Cat** [comment]
Benchmarks added to sympy_benchmarks, #65, will post actual numbers here shortly.

**Comment by Pristine-Cat** [comment]
The benchmarks in three modes defined by `_DOTPRODSIMP_MODE` (I will explain `_DOTPRODSIMP_MODE` when I push the code). False means no dotprodsimp at all (normal master). None means dotprodsimp on for muliply, pow, rref and the determinant functions. True means dotprodsimp for everything where it can be turned on like decompositions and solvers. True doesn't really change a lot - LDL benefits but CH suffers. I will probably include LDL in the default dotprodsimp on group of funcs.

The differences you want to look at are False vs. None. Generally dotprodsimp is faster but in a good number of cases it is slower though remember that in all cases the result is returned already simplified so no extra expensive simplification needs to be done afterwards. I would suggest executing some of these tests manually to see the difference since the sympy benchmarks don't do any `count_ops` to give info on result complexity (especially the cases where you see dotprodsimp significantly slower than master). I would also suggest taking a look at the matrices I used in the benchmark PR and trying some different matrices from those.


```
_DOTPRODSIMP_MODE                     False        None        True

TimeCharPoly.A                      1670.00       32.60       32.90
TimeCharPoly.B                     22900.00      108.00      104.00
TimeCharPoly.C                       failed       28.30       24.90
TimeCharPoly.D                       failed     4770.00     4540.00
TimeCholeskySolve.A                   87.20       79.70      failed
TimeCholeskySolve.B                  158.00      145.00      failed
TimeCholeskySolve.C                  272.00       83.20       91.20
TimeCholeskySolve.D                  409.00      387.00      failed
TimeCoFactor.A                         1.09        6.23        6.02
TimeCoFactor.B                         1.13       20.50       19.40
TimeCoFactor.C                         1.17        2.62        2.62
TimeCoFactor.D                         1.23      968.00      896.00
TimeCoFactorMatrix.A                 300.00      477.00      460.00
TimeCoFactorMatrix.B                1120.00     5410.00     5210.00
TimeCoFactorMatrix.C                3520.00     2910.00     2850.00
TimeCoFactorMatrix.D                1420.00      failed      failed
TimeColumnSpace.A                     62.40       46.70       47.10
TimeColumnSpace.B                    283.00       26.80       27.20
TimeColumnSpace.C                   3180.00        2.04        2.09
TimeColumnSpace.D                    failed      failed      failed
TimeConditionNumber.A                 78.00       89.30       87.20
TimeConditionNumber.B                 66.40       59.50       58.00
TimeConditionNumber.C                  2.17        1.43        1.38
TimeConditionNumber.D                  1.96        1.22        1.20
TimeDetBareiss.A                       8.66        1.30        1.29
TimeDetBareiss.B                      23.10        1.26        1.33
TimeDetBareiss.C                     191.00        3.37        3.35
TimeDetBareiss.D                    3200.00      failed      failed
TimeDetBerkowitz.A                     1.68       24.90       25.60
TimeDetBerkowitz.B                     1.73      100.00       96.40
TimeDetBerkowitz.C                     5.06        4.89        4.79
TimeDetBerkowitz.D                     1.96     2680.00     2510.00
TimeDetLU.A                            1.05        1.36       20.00
TimeDetLU.B                            1.44        1.44        1.69
TimeDetLU.C                            1.92        1.92        2.82
TimeDetLU.D                            1.47        1.44      failed
TimeDiagonalize.A                    175.00      167.00      165.00
TimeDiagonalize.B                    724.00      829.00      763.00
TimeDiagonalize.C                    failed      failed      failed
TimeDiagonalize.D                   4510.00     3900.00     3870.00
TimeEchelonForm.A                     62.50       47.50       47.30
TimeEchelonForm.B                    305.00       27.60       27.50
TimeEchelonForm.C                   3190.00        2.06        2.13
TimeEchelonForm.D                    failed      failed      failed
TimeEigenVals.A                     1660.00       33.90       35.40
TimeEigenVals.B                    22900.00      107.00      116.00
TimeEigenVals.C                      failed      143.00      144.00
TimeEigenVals.D                      failed      failed      failed
TimeEigenVects.A                     600.00      598.00      594.00
TimeEigenVects.B                     523.00      558.00      527.00
TimeEigenVects.C                      73.80       42.20       41.70
TimeEigenVects.D                     962.00     1070.00     1080.00
TimeGaussJordanSolve.A                47.30       49.50       49.40
TimeGaussJordanSolve.B                23.90       31.30       31.40
TimeGaussJordanSolve.C                10.80        1.46        1.47
TimeGaussJordanSolve.D                63.80      368.00      360.00
TimeInvADJ.A                          23.60       22.70       22.80
TimeInvADJ.B                          35.20       30.60       29.30
TimeInvADJ.C                          41.20       17.00       15.30
TimeInvADJ.D                          91.30      204.00      201.00
TimeInvCH.A                          105.00       92.40      failed
TimeInvCH.B                          219.00      181.00      failed
TimeInvCH.C                          467.00      307.00      269.00
TimeInvCH.D                          617.00      592.00      failed
TimeInvGE.A                           48.20       64.90       65.70
TimeInvGE.B                           23.90       50.00       49.70
TimeInvGE.C                           10.80        1.64        1.59
TimeInvGE.D                           67.40      589.00      597.00
TimeInvLDL.A                         failed      failed     1330.00
TimeInvLDL.B                         failed      failed     2440.00
TimeInvLDL.C                         288.00      100.00       98.40
TimeInvLDL.D                         failed      failed      failed
TimeInvLU.A                           23.40       20.80       41.80
TimeInvLU.B                           36.30       22.00       44.60
TimeInvLU.C                           43.50       15.60       15.50
TimeInvLU.D                           93.20      148.00      334.00
TimeIsDiagonalizable.A              1420.00     1020.00     1020.00
TimeIsDiagonalizable.B               748.00      800.00      756.00
TimeIsDiagonalizable.C               116.00       81.00       77.60
TimeIsDiagonalizable.D              1070.00     1200.00     1160.00
TimeIsNilpotent.A                   1660.00       32.80       31.60
TimeIsNilpotent.B                  23100.00      112.00      103.00
TimeIsNilpotent.C                    failed       26.40       25.70
TimeIsNilpotent.D                    failed     4850.00     4670.00
TimeIsPositiveDefiniteCH.A            45.40       47.10       46.90
TimeIsPositiveDefiniteCH.B            34.10       37.10     2930.00
TimeIsPositiveDefiniteCH.C            22.30       24.30       22.70
TimeIsPositiveDefiniteCH.D            32.80       34.60       33.00
TimeIsPositiveDefiniteEigen.A        700.00      570.00      514.00
TimeIsPositiveDefiniteEigen.B        663.00      543.00      537.00
TimeIsPositiveDefiniteEigen.C         62.60       29.80       27.50
TimeIsPositiveDefiniteEigen.D        675.00      825.00      809.00
TimeIsPositiveDefiniteLDL.A           45.90       46.60       48.50
TimeIsPositiveDefiniteLDL.B           58.00       58.00      483.00
TimeIsPositiveDefiniteLDL.C           22.80       23.40       22.60
TimeIsPositiveDefiniteLDL.D           32.50       33.90       33.50
TimeJordanForm.A                    1450.00     1280.00     1270.00
TimeJordanForm.B                     529.00      545.00      542.00
TimeJordanForm.C                     210.00       49.70       44.30
TimeJordanForm.D                    2090.00     2250.00     2310.00
TimeLDLsolve.A                       failed      failed      871.00
TimeLDLsolve.B                       failed      failed     2110.00
TimeLDLsolve.C                       187.00       78.10       80.10
TimeLDLsolve.D                       failed      failed      failed
TimeLUsolve.A                          0.80        0.77       15.20
TimeLUsolve.B                          0.79        0.79       18.60
TimeLUsolve.C                          0.97        1.01        2.08
TimeLUsolve.D                          0.97        0.96      140.00
TimeLeftEigenVects.A                1130.00     1000.00     1010.00
TimeLeftEigenVects.B                 524.00      556.00      550.00
TimeLeftEigenVects.C                  74.70       41.50       41.30
TimeLeftEigenVects.D                1100.00     1310.00     1380.00
TimeNullSpace.A                       62.50       47.50       47.20
TimeNullSpace.B                      298.00       27.60       27.50
TimeNullSpace.C                     3560.00        3.73        3.55
TimeNullSpace.D                      failed      failed      failed
TimePow16.A                            3.11        3.87        3.93
TimePow16.B                          265.00      513.00      543.00
TimePow16.C                         8900.00       54.80       55.90
TimePow16.D                          272.00      failed      failed
TimePow4.A                             1.06        1.96        1.97
TimePow4.B                             0.61       79.10       83.00
TimePow4.C                             1.82        3.13        3.23
TimePow4.D                             0.73     2220.00     2290.00
TimeRREF.A                            63.10       46.40       46.40
TimeRREF.B                           298.00       26.70       26.90
TimeRREF.C                          3540.00        3.36        3.37
TimeRREF.D                           failed      failed      failed
TimeRank.A                            63.30       46.30       46.70
TimeRank.B                           289.00       26.90       27.20
TimeRank.C                          3250.00        2.35        2.37
TimeRank.D                           failed      failed      failed
TimeRowSpace.A                        64.60       46.00       46.50
TimeRowSpace.B                       298.00       27.60       26.80
TimeRowSpace.C                      3310.00        2.25        2.07
TimeRowSpace.D                       failed      failed      failed
TimeSingularValues.A                 507.00      302.00      298.00
TimeSingularValues.B                1030.00      791.00      785.00
TimeSingularValues.C                 462.00        6.36        7.90
TimeSingularValues.D                 673.00     1100.00     1060.00
```

**Comment by Pristine-Cat** [comment]
Turned on dotprodsimp in determinants and added dotprodsimp control variable for emergencies. `_DOTPRODSIMP_MODE` can be `None` for dotprodsimp on only in `multiply`, `pow`, `rref` and the determinants, `False` for dotprodsimp off everywhere and behavior consistent with previous master and `True` for dotprodsimp on everywhere (including decompositions and solvers). There is no user interface to set this currently - though you could add that if you want, and maybe it is even a good idea to add an undocumented method to turn dotprodsimp off in case it breaks user code in something important.

Still have docs to fix and have not yet removed dotprodsimp from func sigs since that will be an ugly diff and complicate you guys reviewing this - so saving for last like I said. The current question I am thinking is if I should turn dotprodsimp default on for LDL and upper/lower triangular solve where it might benefit but I am leaning towards not doing that since we already have enough benefit from where it is on currently.

I had to comment out another test but if you look at that test I think you will understand, the test is `matrices/tests/test_matrices.py:test_pinv` and it is the last part of this test.

**Comment by Pristine-Cat** [comment]
I did actually check to make sure the new results are equivalent to the old ones.

**Comment by Pristine-Cat** [comment]
Since this is kind of an important decision I did the count_ops comparison between dotprodsimp default and no dotprodsimp, updated in the matrix above.

**Comment by oscarbenjamin** [comment]
It looks like charpoly gets much better but detberkowitz gets much worse. I thought they both just called berkowitz_vector...

**Comment by Pristine-Cat** [comment]
> It looks like charpoly gets much better but detberkowitz gets much worse. I thought they both just called berkowitz_vector...

Yes, but you don't get something for free, look at the count_ops reduction.

**Comment by Pristine-Cat** [comment]
Ok, down to last test error but not sure how to handle this one. In `doc/src/modules/physics/mechanics/examples/bicycle_example.rst` line 195 hangs:
```julia
  >>> (fr, frstar) = KM.kanes_equations(BL, FL)
```

**Comment by oscarbenjamin** [comment]
> bicycle_example.rst

This hangs computing a matrix vector product where the matrix is enormous but the vector is trivial. When I interrupt it it always seems to be in cancel called from dotprodsimp.

**Comment by oscarbenjamin** [comment]
Maybe there should be an upper limit on `count_ops`:
```
> sympy/physics/mechanics/kane.py(214)_initialize_constraint_matrices()
(Pdb) p self._k_nh.diff(dynamicsymbols._t).applyfunc(count_ops)
Matrix([
[   0, 0, 1114, 2764, 1676, 330],
[3015, 0, 1637, 2761, 1688, 320],
[2764, 0, 1350,    0, 1687,   0]])
```

**Comment by oscarbenjamin** [comment]
> Its not the count_ops but the actual content which decides hang or no hang.

Maybe having dotprodsimp by default in multiply is too much.

**Comment by Pristine-Cat** [comment]
> Maybe having dotprodsimp by default in multiply is too much.

Actually its the most important place for it to be, the base of everything else (berkowitz does its magic only because of multiply).

**Comment by Pristine-Cat** [comment]
I will try something but am not very optimistic...

**Comment by oscarbenjamin** [comment]
> (berkowitz does its magic only because of multiply).

Can berkowitz not just use `multiply(dotprodsimp=True)`?

I feel like this is approaching from the wrong direction by enabling dotprodsimp by default everywhere. My suggestion was to do it in rref and then re-evaluate the situation before changing anything else.

**Comment by Pristine-Cat** [comment]
> Can berkowitz not just use `multiply(dotprodsimp=True)`?

That is exactly what I am timing and counting ops now.

**Comment by Pristine-Cat** [comment]
Ok, not too bad, it worked. `multiply` defaults to no dotprodsimp but added explicit `multiply(..., dotprodsimp=True)` to `berkowitz`, `exp`, and `jordan_form` to keep the benefits there. The difference from default mul is small, a very few functions gained some complexity or slowed down but in general it is more than acceptable. That doc block now runs and maybe a few others will as well, will check. Updated times and count_ops in matrix above

**Comment by oscarbenjamin** [comment]
Merge conflict...

**Comment by Pristine-Cat** [comment]
All but one of the tests which were hanging before work again now. The remaining test which hangs is the geometry one `geometry/tests/test_line:test_intersection_2d`.

**Comment by Pristine-Cat** [comment]
Finally seems stable. I've gone ahead and removed the `dotprodsimp` arg from all functions. `multiply` remains as the only function which takes this arg and defaults to `False`. `dotprodsimp` multiplication is used in the following functions (either through `multiply(..., dotprodsimp=True)` or a post-multiply `dotprodsimp` simplification directly):
```
_eval_pow_by_recursion_dotprodsimp
_berkowitz_toeplitz_matrix
_berkowitz_vector
_det
_det_bareiss
_jordan_form
MatrixBase.exp
_row_reduce_list
```
So basically determinant/charpoly, rref, jordan_form and exp. I've also reverted any tests which did not need changing for this use of dotprodsimp to their previous state. There is one test which hangs now which did not before and that is the last test of `geometry/tests/test_line:test_intersection_2d`.

I've also put in an emergency shutoff in `utilities.py` in the form of:
```
_DOTPRODSIMP_MODE = False if os.environ.get('SYMPY_DOTPRODSIMP', '').lower() in \
        ('false', 'off', '0') else None
```
So you can revert to previous behavior by setting the environment variable `SYMPY_DOTPRODSIMP` to an off value like `False`, `Off` or `0`. If you don't want this I can remove it but I think having some way of disabling dotprodsimp is useful since it can drastically change the structure of results. In any case if you guys accept this and merge to master I think any problems would make themselves apparent to contributors before this ever goes live.

If you want a commit squash and rebase on current master let me know (assuming all tests pass like they did locally).

**Comment by sylee957** [comment]
Okay, I see you are keeping a good work here and the benchmark result seems reasonable.
Some may have degraded, but in most cases, I see there are more improvements than some degradations.

The only thing that slows down is
`+     2.26±0.02ms         93.0±2ms    41.19  matrices.TimeMatrixPower.time_Case1`
Probably because you're expanding the elements.
But I think that such cases can be solved via other ways, like decomposing a common scalar out of the matrix.

**Comment by Pristine-Cat** [comment]
> The only thing that slows down is
> `+ 2.26±0.02ms 93.0±2ms 41.19 matrices.TimeMatrixPower.time_Case1`
> Probably because you're expanding the elements.

In that case it is because the expression gets large enough in ops for an attempt at `cancel`. I think there is room to optimize in the actual `dotprodsimp` function to detect maybe in the `count_ops_alg` function whether `cancel` would do anything, maybe by checking for negative exponents in powers to see if rational functions are present and if they are not then do not attempt `cancel`.

**Comment by Pristine-Cat** [comment]
@sylee957 inspired me to try something, lets see how the benchmark does...

**Comment by Pristine-Cat** [comment]
It worked:
```
+     2.32±0.03ms      7.95±0.06ms     3.43  matrices.TimeMatrixPower.time_Case1
```
Now I just need to confirm with my own tests that it didn't break the simplification.

**Comment by Pristine-Cat** [comment]
Ok well that was totally worth it. The simplification is untouched but in some cases the speedup is very nice. I've updated the performance matrix at the top of this PR, you can see the ops are the same but in some cases the speed is significantly better, for example:
```julia
CharPoly.B   103ms ->   11ms
CharPoly.D  4860ms -> 3010ms
CoFactor.D   893ms ->  357ms
Pow16.B      523ms ->   91ms
```
Unless there are further comments / checks / requests I would say this PR is done.

**Comment by oscarbenjamin** [comment]
This looks good!

I don't think that the environment variable is necessary but I won't kick up a fuss...

**Comment by Pristine-Cat** [comment]
Added three tiny changes which didn't make it into this PR in #18572, they are not terribly important but very easy to check.

**Comment by oscarbenjamin** [comment]
This has lead to a performance regression in matrix power with a complicated matrix:
```
import time

def makeMF(n):
    from pydy.models import n_link_pendulum_on_cart
    sys = n_link_pendulum_on_cart(n=n)
    M = sys.eom_method.mass_matrix
    F = sys.eom_method.forcing
    return M, F

n = 9
M, F = makeMF(n)
start = time.time()
M**2  # matrix power has slowed down
delta = time.time() - start
print('n=%d time=%.2e secs' % (n, delta))
```
On current master `M**2` takes 20 seconds whereas in sympy 1.5 it takes 0.1 seconds.

File: sympy/matrices/utilities.py [hunk 1]
```diff
@@ -1,6 +1,56 @@
 from __future__ import division, print_function
 
+import os
+
 from sympy.core.function import expand_mul
+from sympy.simplify.simplify import dotprodsimp as _dotprodsimp
+
+
+# The following is an internal variable for controlling the recently introduced
```
Comment (asmeurer) [inline sympy/matrices/utilities.py:9]: The word "recently" quickly becomes incorrect in a comment.

**Comment by asmeurer** [comment]
I also don't like the environment variable. I think we should minimize the number of configuration options we have outside of keyword arguments. Also environment variables in general are a poor way to configure something like this because they can only be set once before SymPy is imported. For global things like the cache that can't be changed after SymPy is imported, this makes sense, but I don't understand why this option should be like that. Why can't it just be a keyword argument?

**Pull request merged**